lpc55_sign/src/signed_image.rs

```diff
@@ -24,366 +36,166 @@ fn get_pad(val: usize) -> usize {
+    // of each certificate.
+    let mut cert_table = Vec::new();
+    for cert in &signing_certs {
+        let cert_pad = get_pad(cert.len());
+        let padded_len = cert.len() + cert_pad;
+        cert_table.extend_from_slice(&(padded_len as u32).to_le_bytes());
+        cert_table.extend_from_slice(cert);
+        if cert_pad > 0 {
+            cert_table.extend_from_slice(&vec![0; cert_pad]);
+        }
+    }
+    let cert_table_len = cert_table.len();
+    let cert_header_len = CertHeader::packed_bytes_size(None)?;
+    let mut cert_header: CertHeader = CertHeader::new(cert_header_len, cert_table_len);
+    cert_header.certificate_count = signing_certs.len() as u32;
+    let cert_table_len = cert_header.certificate_table_len as usize;
+
+    // How many bytes we sign, including image, cert table, and root key hashes.
+    let image_len = image_bytes.len();
+    let image_pad = get_pad(image_len);
+    let signed_len = image_len + image_pad + cert_header_len + cert_table_len + 4 * 32;
+    cert_header.total_image_len = signed_len.try_into()?;
+
+    // Total image length includes the length of the eventual signature.
+    let (_, leaf) = parse_x509_certificate(&signing_certs.last().unwrap())?;
+    let pub_key = RsaPublicKey::from_pkcs1_der(leaf.public_key().subject_public_key.as_ref())?;
+    let sig_len = pub_key.n().bits() / 8;
+    let total_len = signed_len + sig_len;
+
+    // Start writing the image header: first the total image length.
+    LittleEndian::write_u32(&mut image_bytes[0x20..0x24], total_len as u32);
+
+    // Next comes the boot field.
     let boot_field = BootField::new(BootImageType::SignedImage);
+    image_bytes[0x24..0x28].clone_from_slice(&boot_field.pack()?);
 
-    bytes[0x24..0x28].clone_from_slice(&boot_field.pack()?);
-    // Our execution address goes in the next word
-    byteorder::LittleEndian::write_u32(&mut bytes[0x34..0x38], execution_address);
-    // where to find the block. For now just stick it right after the image
-    byteorder::LittleEndian::write_u32(&mut bytes[0x28..0x2c], (image_len + image_pad) as u32);
-
-    let mut out = OpenOptions::new()
-        .write(true)
-        .truncate(true)
-        .append(false)
-        .create(true)
-        .open(outfile_path)?;
+    // Then the location of the certificate table: right after the image.
+    LittleEndian::write_u32(&mut image_bytes[0x28..0x2c], (image_len + image_pad) as u32);
 
-    // Need to write out an empty sha since we only have one root key
-    let empty_hash: [u8; 32] = [0; 32];
+    // Optionally write the image execution address.
+    if execution_address > 0 {
+        LittleEndian::write_u32(&mut image_bytes[0x34..0x38], execution_address);
+    }
 
-    out.write_all(&bytes)?;
+    // Generate the image, see 7.3.4 of v2.4 UM 11126 for the layout.
     if image_pad > 0 {
-        out.write_all(&vec![0; image_pad])?;
+        image_bytes.extend_from_slice(&vec![0; image_pad])
```

Comment:
```suggestion
        image_bytes.resize(image_bytes.len() + image_pad, 0);
```
(and in fact, the conditional is not necessary if you write it like this)

Reply:
Indeed, thanks! Applied in b3520b7.